smartsim/_core/entrypoints/telemetrymonitor.py

```diff
@@ -0,0 +1,688 @@
+    )
+    return manifest
+
+
+def track_event(
+    timestamp: int,
+    task_id: t.Union[int, str],
+    step_id: str,
+    etype: str,
+    action: _EventClass,
+    status_dir: pathlib.Path,
+    logger: logging.Logger,
+    detail: str = "",
+    return_code: t.Optional[int] = None,
+) -> None:
+    """
+    Persist a tracking event for an entity
+    """
+    tgt_path = status_dir / f"{action}.json"
+    tgt_path.parent.mkdir(parents=True, exist_ok=True)
+
+    try:
+        task_id = int(task_id)
+    except ValueError:
+        pass
+
+    entity_dict = {
+        "timestamp": timestamp,
+        "job_id": task_id,
+        "step_id": step_id,
+        "type": etype,
+        "action": action,
+    }
+
+    if detail is not None:
+        entity_dict["detail"] = detail
+
+    if return_code is not None:
+        entity_dict["return_code"] = return_code
+
+    try:
+        if not tgt_path.exists():
+            """Don't overwrite existing tracking files"""
+            bytes_written = tgt_path.write_text(json.dumps(entity_dict, indent=2))
+            if bytes_written < 1:
+                logger.warning("event tracking failed to write tracking file.")
+    except Exception:
+        logger.error("Unable to write tracking file.", exc_info=True)
+
+
+def track_timestep(job: Job, logger: logging.Logger) -> None:
+    """Persists telemetry event for a timestep"""
+    if hasattr(job.entity, "status_dir"):
+        write_path = pathlib.Path(job.entity.status_dir)
+    else:
+        write_path = pathlib.Path(job.entity.path)
+
+    track_event(
+        get_ts(),
+        job.jid or "" if not job.is_task else "",
+        job.jid or "" if job.is_task else "",
+        job.entity.type,
+        "timestep",
+        write_path,
+        logger,
+    )
+
+
+class ManifestEventHandler(PatternMatchingEventHandler):
+    """The ManifestEventHandler monitors an experiment for changes and updates
+    a telemetry datastore as needed.
+
+    It contains event handlers that are triggered by changes to a runtime experiment
+    manifest. The runtime manifest differs from a standard manifest. A runtime manifest
+    may contain multiple experiment executions in a `runs` collection.
+
+    It also contains a long-polling loop that checks experiment entities for updates
+    at each timestep."""
+
+    def __init__(
+        self,
+        pattern: str,
+        logger: logging.Logger,
+        ignore_patterns: t.Any = None,
+        ignore_directories: bool = True,
+        case_sensitive: bool = False,
+    ) -> None:
+        super().__init__(
+            [pattern], ignore_patterns, ignore_directories, case_sensitive
+        )  # type: ignore
+        self._logger = logger
+        self._tracked_runs: t.Dict[int, Run] = {}
+        self._tracked_jobs: t.Dict[_JobKey, JobEntity] = {}
+        self._completed_jobs: t.Dict[_JobKey, JobEntity] = {}
+        self._launcher_type: str = ""
+        self._launcher: t.Optional[Launcher] = None
```

Comment:
Not wrong, but do we strictly need both a launcher and a launcher type? Do we need to know the launcher type before setting the launcher itself?

Reply:
removed